### PR TITLE
PTX-24561 To Test App in 3 diff combinations of Realm-pod,pod,and legacy deployment

### PR DESCRIPTION
**Pull request by sayalasomayajula-px**
…in APP_LIST - realm::pod, pod, usual volume (no realm, no pod)

**What this PR does / why we need it**:
PTX-24561

**Which issue(s) this PR fixes** (optional)
Closes #
In cases where pure.json has combination of arrays with realm and without realm, we need a test which deploys FADA applications in below combinations (for every APP provided in APP_LIST)

Storage Class having pure_fa_pod_name - this pod is under a realm in one of the arrays in pure.json

Storage Class having pure_fa_pod_name - this pod is not under a realm in one of the arrays in pure.json

Storage Class not having pure_fa_pod_name - legacy way of FADA Apps

So in essence, if we pass one app (ex: fio-fa-davol) we should be creating 3 APPs (fio-fa-davol-realm-pod, fio-fa-davol-pod, fio-fa-davol)

**Special notes for your reviewer**:
PASS LOGS:https://jenkins.pwx.dev.purestorage.com/blue/organizations/jenkins/Users%2Fkrishna%2Ftp-pxe-byoc/detail/tp-pxe-byoc/3447/pipeline

### CODE REVIEW

File: tests/basic/pure_test.go [hunk 1]
```diff
@@ -5130,3 +5130,125 @@ var _ = Describe("{ValidatePodNameinVolume}", func() {
 		AfterEachTest(contexts)
 	})
 })
+var _ = Describe("{MultiTenancyFATestWithPodRealm}", func() {
```
Comment (vumesh-px) [inline tests/basic/pure_test.go:5133]: we can rename this more appropriately

```suggestion
var _ = Describe("{FAMultiTenancyMultiAppWithPodRealm}", func() {
```

File: tests/basic/pure_test.go [hunk 2]
```diff
@@ -5130,3 +5130,125 @@ var _ = Describe("{ValidatePodNameinVolume}", func() {
...
+		case faWithoutRealm == nil:
+			log.FailOnError(fmt.Errorf("No  FA without realm found in pure.json"), "No  FA without realm found in pure.json")
+		case !isRealmFAAccessible:
+			log.FailOnError(fmt.Errorf("No accessible realm FA found in pure.json"), "No accessible realm FA found in pure.json")
+		}
+		podCreateandAppDeploy := func(faclient *newFlashArray.Client, podNameinFA string, podNameinSC string, taskName string, isMultiTenancy bool) {
+			var contexts []*scheduler.Context
+			var pvcList []string
+			if isMultiTenancy {
+				_, err = pureutils.CreatePodinFA(faclient, podNameinFA)
```
Comment (vumesh-px) [inline tests/basic/pure_test.go:5198]: check for existence before creating

File: tests/basic/pure_test.go [hunk 3]
```diff
@@ -5130,3 +5130,125 @@ var _ = Describe("{ValidatePodNameinVolume}", func() {
...
+			faWithoutRealm      *newFlashArray.Client
+			isRealmFAAccessible bool
+			podNameinSC         string
+			PodNameinFA         string
+		)
+		applist := Inst().AppList
+		defer func() {
+			Inst().AppList = applist
+		}()
+		Inst().AppList = []string{"fio-fa-davol"}
```
Comment (vumesh-px) [inline tests/basic/pure_test.go:5167]: I believe we can skip hard-coding and assert if applist is nil

File: tests/basic/pure_test.go [hunk 4]
```diff
@@ -5130,3 +5130,125 @@ var _ = Describe("{ValidatePodNameinVolume}", func() {
...
+
+		switch {
+		case faWithRealm == nil:
+			log.FailOnError(fmt.Errorf("No FA with realm found in pure.json"), "No  FA with realm found in pure.json")
+		case faWithoutRealm == nil:
+			log.FailOnError(fmt.Errorf("No  FA without realm found in pure.json"), "No  FA without realm found in pure.json")
+		case !isRealmFAAccessible:
+			log.FailOnError(fmt.Errorf("No accessible realm FA found in pure.json"), "No accessible realm FA found in pure.json")
+		}
+		podCreateandAppDeploy := func(faclient *newFlashArray.Client, podNameinFA string, podNameinSC string, taskName string, isMultiTenancy bool) {
```
Comment (vumesh-px) [inline tests/basic/pure_test.go:5194]: podNameinSC could the return value, I don't see a value in taking it in as an argument

File: tests/basic/pure_test.go [hunk 5]
```diff
@@ -5130,3 +5130,125 @@ var _ = Describe("{ValidatePodNameinVolume}", func() {
...
+			log.FailOnError(faErr, "Failed to check if volumes created  exist in FA")
+
+		}
+		stepLog = "Deploy FADA application with storage class having pure_fa_pod_name - this pod is under a realm in one of the arrays in pure.json"
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			podNameinSC = "Torpedo-Test" + time.Now().Format("01-02-15h04m05s")
+			// In order to create a pod inside a realm , the pod name should be prefixed with realm, eg: <realm name>::<pod name>
+			PodNameinFA = realmName + "::" + podNameinSC
+			podCreateandAppDeploy(faWithRealm, PodNameinFA, podNameinSC, "fada-app-with-pod-realm", true)
```
Comment (vumesh-px) [inline tests/basic/pure_test.go:5235]: take in only podName as an argument. Evaluate the podNameinFA in the func by validating the realmName (from the provided `faWithRealm`)

File: tests/basic/pure_test.go [hunk 6]
```diff
@@ -5130,3 +5130,125 @@ var _ = Describe("{ValidatePodNameinVolume}", func() {
...
+		})
+		stepLog = "Deploy FADA application with storage class not having pure_fa_pod_name"
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			// Here we dont want to create any pod and create pure_fa_pod_name in storage class ,as this is standard way of Deploying Apps
+			podCreateandAppDeploy(faWithRealm, "", "", "fada-app-without-pod-realm", false)
```
Comment (vumesh-px) [inline tests/basic/pure_test.go:5241]: this would fail on `faWithRealm`

File: tests/basic/pure_test.go [hunk 7]
```diff
@@ -5130,3 +5130,125 @@ var _ = Describe("{ValidatePodNameinVolume}", func() {
...
+			log.InfoD(stepLog)
+			podNameinSC = "Torpedo-Test" + time.Now().Format("01-02-15h04m05s")
+			// pod name will not contain any realm , so we just create pod with same naming convention in FA
+			PodNameinFA = podNameinSC
+			podCreateandAppDeploy(faWithoutRealm, PodNameinFA, podNameinSC, "fada-app-with-pod", true)
+		})
+
+	})
+
+})
```
Comment (vumesh-px) [inline tests/basic/pure_test.go:5254]: can we also enhance this test with more combinations?
on faWithRealm:
1. create pod without realm -- negative case
2. create pod under realm
3. create without pure_fa_pod_name

on faWithoutRealm:
1. create pod (no realm)
2. create pod under an existing realm (random) -- negative case
2. create without pure_fa_pod_name

**Review by vumesh-px — CHANGES_REQUESTED** [review]

File: tests/basic/pure_test.go (hunk 7 above)
Comment (vumesh-px) [reply tests/basic/pure_test.go:5254]: On further analysis with @sayalasomayajula-px ,
on faWithRealm:

create pod without realm -- negative case, need to add
create pod under realm -- already covered 
create without pure_fa_pod_name -- will be covered as part of existing fa regression jobs
on faWithoutRealm:

create pod (no realm) -- @Pure-AdamuKaapan can you confirm if FADA PVC with just pod name (without realm) is already covered?
create pod under an existing realm (random) -- negative case
create without pure_fa_pod_name -- will be covered as part of existing fa regression jobs

File: tests/basic/pure_test.go (hunk 2 above)
Comment (sayalasomayajula-px) [reply tests/basic/pure_test.go]: added

File: tests/basic/pure_test.go (hunk 3 above)
Comment (sayalasomayajula-px) [reply tests/basic/pure_test.go]: Thanks, Removed it

File: tests/basic/pure_test.go (hunk 5 above)
Comment (sayalasomayajula-px) [reply tests/basic/pure_test.go:5229]: As per discussion , leaving it like this

File: tests/basic/pure_test.go [hunk 8]
```diff
@@ -5130,3 +5130,152 @@ var _ = Describe("{ValidatePodNameinVolume}", func() {
...
+			if wrongPodoutSideRealm {
+				fmt.Println("Entering into loop to check if it entered negative scenario")
+				for _, ctx := range contexts {
+					ctx.SkipVolumeValidation = true
+					log.InfoD("waiting for a minute for volume name to populate")
+					time.Sleep(1 * time.Minute)
+					allPvcList, err := core.Instance().GetPersistentVolumeClaims(ctx.App.NameSpace, nil)
+					log.FailOnError(err, fmt.Sprintf("error getting pvcs from namespace [%s]", ctx.App.NameSpace))
+					for _, p := range allPvcList.Items {
+						fmt.Println("PVC Name is ", p.Name)
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:5219]: remove fmt.print

File: tests/common.go [hunk 9]
```diff
@@ -13369,3 +13369,31 @@ func DeleteFilesFromS3Bucket(bucketName string, fileName string) error {
...
+	if err != nil {
+		log.InfoD("error getting pvcs from namespace [%s]", namespace)
+		return nil, err
+	}
+
+	for _, p := range allPvcList.Items {
+		pvclist = append(pvclist, p.Name)
+		return pvclist, nil
+	}
+	return nil, nil
```
Comment (lsrinivas-pure) [inline tests/common.go:13384]: return error here no pvc found in the given namespace

File: tests/common.go [hunk 10]
```diff
@@ -13369,3 +13369,31 @@ func DeleteFilesFromS3Bucket(bucketName string, fileName string) error {
...
+}
+
+func GetVolumeNamefromPVC(namespace string) ([]string, error) {
```
Comment (lsrinivas-pure) [inline tests/common.go:13387]: logic in this function and above function is same

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/common.go (hunk 10 above)
Comment (sayalasomayajula-px) [reply tests/common.go:13373]: removed it

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Review by vumesh-px — APPROVED** [review]

**Pull request merged**